src/sql-parser/src/parser.rs

```diff
@@ -3943,6 +3947,24 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_rows_from(&mut self) -> Result<TableFactor<Raw>, ParserError> {
+        self.expect_token(&Token::LParen)?;
+        let functions = self.parse_comma_separated(Parser::parse_table_function_no_alias)?;
+        self.expect_token(&Token::RParen)?;
+        let alias = self.parse_optional_table_alias()?;
+        Ok(TableFactor::RowsFrom { functions, alias })
+    }
+
+    fn parse_table_function_no_alias(&mut self) -> Result<TableFunction<Raw>, ParserError> {
```

Comment:
Not blocking: how would you feel about something that didn't require creating a new function to express whether or not to parse an AST element? Just to see if it was tractable, I typed out a diff

```diff
diff --git a/src/sql-parser/src/parser.rs b/src/sql-parser/src/parser.rs
index 07d61416d..84be1f8b4 100644
--- a/src/sql-parser/src/parser.rs
+++ b/src/sql-parser/src/parser.rs
@@ -3842,19 +3842,13 @@ impl<'a> Parser<'a> {
     fn parse_table_factor(&mut self) -> Result<TableFactor<Raw>, ParserError> {
         if self.parse_keyword(LATERAL) {
             // LATERAL must always be followed by a subquery or table function.
-            if self.consume_token(&Token::LParen) {
-                return self.parse_derived_table_factor(Lateral);
+            return if self.consume_token(&Token::LParen) {
+                self.parse_derived_table_factor(Lateral)
             } else if self.parse_keywords(&[ROWS, FROM]) {
-                return Ok(self.parse_rows_from()?);
+                self.parse_rows_from()
             } else {
-                let name = self.parse_object_name()?;
-                self.expect_token(&Token::LParen)?;
-                return Ok(TableFactor::Function(TableFunction {
-                    name,
-                    args: self.parse_optional_args(false)?,
-                    alias: self.parse_optional_table_alias()?,
-                }));
-            }
+                Ok(TableFactor::Function(self.parse_table_function()?))
+            };
         }
 
         if self.consume_token(&Token::LParen) {
@@ -3949,19 +3943,30 @@ impl<'a> Parser<'a> {
 
     fn parse_rows_from(&mut self) -> Result<TableFactor<Raw>, ParserError> {
         self.expect_token(&Token::LParen)?;
-        let functions = self.parse_comma_separated(Parser::parse_table_function_no_alias)?;
+        let functions = self.parse_comma_separated(|parser| {
+            let tf = parser.parse_table_function()?;
+            if tf.alias.is_some() {
+                parser_err!(
+                    parser,
+                    parser.peek_prev_pos(),
+                    "ROWS FROM does not support individual aliases in table functions"
+                )
+            } else {
+                Ok(tf)
+            }
+        })?;
         self.expect_token(&Token::RParen)?;
         let alias = self.parse_optional_table_alias()?;
         Ok(TableFactor::RowsFrom { functions, alias })
     }
 
-    fn parse_table_function_no_alias(&mut self) -> Result<TableFunction<Raw>, ParserError> {
+    fn parse_table_function(&mut self) -> Result<TableFunction<Raw>, ParserError> {
         let name = self.parse_object_name()?;
         self.expect_token(&Token::LParen)?;
         Ok(TableFunction {
             name,
             args: self.parse_optional_args(false)?,
-            alias: None,
+            alias: self.parse_optional_table_alias()?,
         })
     }
```

I realize this is more LOC but the special-cased function used only in one spot feels off to me.

Reply:
:+1: but I feel like we should go one step further and make the AST enforce this with types. ISTM that `TableFunction` should not have an `alias` field, and `TableFactor::Function` should look like:

```
TableFunction::Function {
    function: TableFunction,
    alias: Option<Alias>,
}
```

Reply:
Oh yeah, that is much better

Reply:
Neat!